### PR TITLE
[docs] add info about mismatched keys

### PR DESCRIPTION
**Pull request by Meschreiber**
This PR adds information about "mismatched" `@keys` across subgraphs.

Refer to https://apollograph.slack.com/archives/C03RH11Q6S1/p1685024491858819 for more info.

### CODE REVIEW

File: docs/source/entities-advanced.mdx [hunk 1]
```diff
@@ -57,6 +57,32 @@ type Organization {
...
+type Product @key(fields: "sku") {
+  sku: ID!
+  inStock: Boolean!
+}
+```
+
+</CodeColumns>
+
+As long as `id` is a unique identifier within the products subgraph and `sku` is a unique identifier for the same product in the inventory subgraph, the subgraphs can each resolve the `Product` type.
+In other words, **each key declaration is specific to the subgraph it belongs to**, and you don't need to match `@key` fields between graphs.
```
Comment (lleadbet) [inline docs/source/entities-advanced.mdx:84]: To clarify this a bit, the `key` acts as a sort of sign to the query planner that says "hey you! i need this data to be able to identify this thing!". 

Just like talking to them, you'd also need to have the thing they're asking for in order for it to work. In the above example, the query planner wouldn't work. It'd choke because entering from either products or inventory can't ever lead to the other subgraph. 

This is the crux of _why_ we've historically made it so that the keys must match. 

A concrete example of this working would be: 

sub a:
```gql
type Query {
  pi: Product
}

type Product @key(fields: "sku") {
  sku: ID!
  upc: String!
  inStock: Boolean!
}
```

sub b: 
```gql
type Product @key(fields: "upc") {
  upc: String! @external
  name: String!
  price: Int
}
```

But as soon as you add a query to B, it needs to reference _something_ from sub a. It's a really weird edge-case and isn't going to come up tremendously often. 

@pcmanus may be able to explain this or at least confirm I'm not totally wrong here.

**Review by lleadbet — APPROVED** [review]

File: docs/source/entities.mdx [hunk 2]
```diff
@@ -55,14 +55,14 @@ type Product @key(fields: "id") {
 
 The `@key` directive defines the entity's **primary key**, which consists of one or more of the type's `fields`. In the example above, the `Product` entity's primary key is its `id` field.
 
-**Every instance of an entity must be uniquely identifiable by its `@key` fields.** This is what enables your router to associate field data from _different_ subgraphs with the _same_ entity instance.
```
[the remark this answers is not in the document]
Comment (Meschreiber) [reply docs/source/entities.mdx:58]: I guess it must be true. (Otherwise, how would entities work?)

I'm a little confused though—if the keys can be mismatched, how do the keys help the router associate entity fields across subgraphs?

File: docs/source/entities.mdx [hunk 3]
```diff
@@ -55,14 +55,14 @@ type Product @key(fields: "id") {
...
+**Every instance of an entity must be uniquely identifiable by its `@key` fields within its own subgraph.**
```
Comment (Meschreiber) [inline docs/source/entities.mdx:58]:
```suggestion
**Every instance of an entity must be uniquely identifiable by its `@key` fields within its own subgraph.** This is what enables your router to associate field data from _different_ subgraphs with the _same_ entity instance.
```

File: docs/source/entities.mdx (hunk 2 above)
[the remark this answers is not in the document]
Comment (lleadbet) [reply docs/source/entities.mdx:58]: The other message describes it better (I think at least)- it's a weird thing.

File: docs/source/entities-advanced.mdx (hunk 1 above)
Comment (dariuszkuc) [reply docs/source/entities-advanced.mdx:84]: I guess the most common use case for different `@key`s is that in one subgraph you could have multiple `@key`s and in another subgraph just one of them. 

While technically subgraphs don't have to share the same `@key`, practically 
- if they share the same fields (e.g. `upc` in the example above, which doesn't need the `@external`) - why wouldn't you define it as a `@key` in all subgraphs that have that field (which gives more flexibility to query planner)?
- if they don't share any of the fields - you won't be able to cross subgraph boundaries, i.e. you won't be able to merge entities as there is no way to merge entities using information from just one subgraph. **Queries against single subgraph will work though**

Comment (Meschreiber) [reply docs/source/entities-advanced.mdx]: This is very helpful, thank you @lleadbet and @dariuszkuc ! Just updated the text based on this info, would you take another look whenever you get a chance?

File: docs/source/entities-advanced.mdx [hunk 4]
```diff
@@ -57,6 +57,99 @@ type Organization {
...
+```graphql title="Inventory subgraph"
+type Product @key(fields: "upc") {
+  upc: String!
+  inStock: Boolean!
+}
+```
+
+</CodeColumns>
+
+The following would allow you to merge the `Product` entity since both subgraph schemas include the `upc` field, even if it's only the `@key` for one of them:
```
Comment (dariuszkuc) [inline docs/source/entities-advanced.mdx:107]: well in this example it will be possible to resolve entity when starting from `products` subgraph but you won't be able to federate calls coming from `inventory` (as `upc` is not a `@key`)

File: docs/source/entities-advanced.mdx [hunk 5]
```diff
@@ -57,6 +57,99 @@ type Organization {
...
+```graphql title="Inventory subgraph"
+type Product @key(fields: "upc") {
+  upc: String!
+  inStock: Boolean!
+}
+```
+
+</CodeColumns>
+
+If subgraphs share any fields—for example `upc` in the preceding example—it's best to use these shared field(s) as `@keys` for optimal [query planner](./entities#the-query-plan) performance.
```
Comment (dariuszkuc) [inline docs/source/entities-advanced.mdx:129]: subgraphs can share number of fields (for optimization purposes -> e.g. this is a common use case behind `@provides`) which cannot be used as keys as they do not uniquely identify the entity. Maybe rephrase it to indicate that if subgraphs share any fields that can be used to uniquely identify underlying entity, then they could be included as `@keys`. But again if there is a "better" key (e.g. single field vs composite key from multiple keys) they may never be used so...

File: docs/source/entities-advanced.mdx [hunk 6]
```diff
@@ -57,6 +57,99 @@ type Organization {
...
+```graphql title="Inventory subgraph"
+type Product @key(fields: "upc") {
+  upc: String!
+  inStock: Boolean!
+}
+```
+
+</CodeColumns>
+
+To merge entities between subgraphs, the entity must have at least one shared field between subgraphs. For example, you can't merge the `Product` entity defined in the following subgraphs because they don't share any fields:
```
Comment (dariuszkuc) [inline docs/source/entities-advanced.mdx:84]: maybe something along those lines
```suggestion
To merge entities between subgraphs, the entity must have at least one shared field between subgraphs. For example, you can't merge the `Product` entity defined in the following subgraphs because they don't share any fields specified in `@key` selection set:
```

File: docs/source/entities-advanced.mdx (hunk 4 above)
Comment (Meschreiber) [reply docs/source/entities-advanced.mdx:107]: Ah, got it. It seems this is maybe an unrealistic example and this should just be omitted :)

File: docs/source/entities-advanced.mdx (hunk 5 above)
Comment (Meschreiber) [reply docs/source/entities-advanced.mdx:129]: > if there is a "better" key (e.g. single field vs composite key from multiple keys) they may never be used so...

What is best for query planner optimization? Just to use one single key field as long as its available on all subgraphs? Or is allowing for multiple keys better?

Comment (dariuszkuc) [reply docs/source/entities-advanced.mdx]: based on https://github.com/apollographql/federation/issues/2695 it appears that we prefer the simpler `@key`s

File: docs/source/entities-advanced.mdx [hunk 7]
```diff
@@ -57,6 +57,77 @@ type Organization {
...
+```
+
+```graphql title="Inventory subgraph"
+type Product @key(fields: "upc") {
+  upc: String!
+  inStock: Boolean!
+}
+```
+
+</CodeColumns>
```
Comment (dariuszkuc) [inline docs/source/entities-advanced.mdx:129]: this section now seems redundant as thats what the `@key`s are for
```suggestion
```

Comment (Meschreiber) [reply docs/source/entities-advanced.mdx:129]: Agreed!

File: docs/source/entities-advanced.mdx (hunk 5 above)
Comment (Meschreiber) [reply docs/source/entities-advanced.mdx]: Got it, thank you!

**Review by smyrick — APPROVED** [review]

**Review by dariuszkuc — APPROVED** [review]
Approved but just wondering whether we should call out explicitly the implication of having mismatched keys -> entities can be expanded as long as there exists a traversable path from the root query to access the fields

for example given the following schema
```graphql
// products subgraph
type Product @key(fields: "sku") {
  sku: ID!
  upc: String!
  name: String!
  price: Int
}

type Query {
  product(sku: ID!): Product
  products: [Product!]!
}

// inventory subgraph
type Product @key(fields: "upc") {
  upc: String!
  inStock: Boolean!
}
```

It will always be possible to resolve all entity fields through all queries as we can join entities through `upc` field. On the other hand if we add a query to the inventory subgraph

```graphql
type Product @key(fields: "upc") {
  upc: String!
  inStock: Boolean!
}

// new query
type Query {
  productsInStock: [Product!]!
}
```

not all queries will be able to federate the calls -> queries from product subgraph will be able to resolve `inStock` field from inventory subgraph BUT `productsInStock` query will not be able to federate the call as there is no `@key(fields: "upc")` in product subgraph and we don't have `sku` field.

**Review by lleadbet — APPROVED** [review]
